README.md
<div align="center">
    <img alt="Pass Butler – Private Cloud Password Manager" src="./projectcover.png" width="600">
</div>

# Pass Butler synchronization server

The Pass Butler server provides the possibility to synchronize passwords between multiple devices.

## Production setup on Debian

The following steps are tested with Debian 11.

### Installation

Add APT repository:

    $ echo "deb https://apt.passbutler.de bullseye main" | sudo tee /etc/apt/sources.list.d/passbutler.list

Add APT repository signing key and update package index:

    $ wget -q -O - https://apt.passbutler.de/signing-key.gpg | sudo apt-key add -
    $ sudo apt update

Install the package:

    $ sudo apt install passbutler-server

### Deployment with Gunicorn and Nginx

Install the package:

    $ sudo apt install gunicorn

Add dedicated user and group:

    $ sudo adduser --system --group --disabled-password --home /var/lib/passbutler-server/example/ "passbutler-server-example"

Create log file and correct owner:

    $ sudo touch /var/log/passbutler-server-example.log
    $ sudo chown passbutler-server-example:passbutler-server-example /var/log/passbutler-server-example.log

Create configuration file `/etc/passbutler-server/example.conf`:

    DATABASE_FILE = '/var/lib/passbutler-server/example/database.sqlite'
    LOG_FILE = '/var/log/passbutler-server-example.log'

    SECRET_KEY = 'SECRET-KEY-PLACEHOLDER'

    ENABLE_REQUEST_LOGGING = False

    REGISTRATION_ENABLED = True
    REGISTRATION_INVITATION_CODE = 'REGISTRATION-INVITATION-CODE-PLACEHOLDER'

Apply random secret and invitation code:

    $ sudo sed -i "s/SECRET-KEY-PLACEHOLDER/$(cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 64 | head -n 1)/g" /etc/passbutler-server/example.conf
    $ sudo sed -i "s/REGISTRATION-INVITATION-CODE-PLACEHOLDER/$(cat /dev/urandom | tr -dc 'A-Z0-9' | fold -w 4 | head -n 4 | paste -sd '-')/g" /etc/passbutler-server/example.conf

Create Systemd service file `/etc/systemd/system/passbutler-server-example.service`:

    [Unit]
    Description=Pass Butler synchronization server daemon (example)
    Requires=passbutler-server-example.socket
    After=network.target

    [Service]
    PIDFile=/run/passbutler-server-example/pid
    User=passbutler-server-example
    Group=passbutler-server-example
    RuntimeDirectory=passbutler-server-example
    WorkingDirectory=/var/lib/passbutler-server/example/
    Environment="PASSBUTLER_SETTINGS=/etc/passbutler-server/example.conf"
    ExecStart=/usr/bin/gunicorn --name=gunicorn-passbutler-server-example --pid /run/passbutler-server-example/pid --workers=1 --pythonpath=/opt/venvs/passbutler-server/lib/python3.8/site-packages --bind=unix:/run/passbutler-server-example/socket.sock 'passbutlerserver:createApp()'
    ExecReload=/bin/kill -s HUP $MAINPID
    ExecStop=/bin/kill -s TERM $MAINPID
    PrivateTmp=true

    [Install]
    WantedBy=multi-user.target

Create Systemd socket file `/etc/systemd/system/passbutler-server-example.socket`:

    [Unit]
    Description=Pass Butler synchronization server socket (example)

    [Socket]
    ListenStream=/run/passbutler-server-example/socket.sock

    [Install]
    WantedBy=sockets.target

Enable Systemd service:

    $ sudo systemctl daemon-reload
    $ sudo systemctl start passbutler-server-example.socket
    $ sudo systemctl enable passbutler-server-example.socket

The socket will be available automatically if it is requested by Nginx later.

Add to your Nginx `sites-available/example.vhost` configuration:

    server {
        ...

        location / {
            proxy_pass http://unix:/run/passbutler-server-example/socket.sock;

            include "proxy_params";
        }
    }

Restart Nginx:

    $ sudo systemctl restart nginx.service

## Development setup

The following steps are tested with Ubuntu 20.04.

Install package:

    $ sudo apt install python3-virtualenv

Setup virtual environment:

    $ virtualenv ~/Desktop/passbutler-server-venv --python=python3
    $ source ~/Desktop/passbutler-server-venv/bin/activate

Clone repository:

    $ git clone ssh://git@github.com/passbutler/passbutler-server.git
    $ cd ./passbutler-server/

Install dependencies:

    $ pip3 install -r requirements.txt

Install dependencies for testing:

    $ pip3 install -r dev-requirements.txt

### Unit testing

Run unit tests (add `-v` for more verbosity and `-s` if you want to see `print` logs also for successful tests):

    $ pytest

### Development server

Start server:

    $ PASSBUTLER_SETTINGS=./passbutlerserver-example.conf flask --app='passbutlerserver:createApp' --debug run --host 0.0.0.0 --port 5000

### Packaging

Currently, only the distribution with "deb" package is supported!

#### Package for Debian/Ubuntu

Install package:

    $ sudo apt install devscripts

Build the package:

    $ debuild

The "deb" file is created in the parent directory.

## License

Pass Butler is licensed under the GNU Affero General Public License 3:

    Copyright (c) 2019-2022 Bastian Raschke

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License, version 3,
    as published by the Free Software Foundation.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program. If not, see <https://www.gnu.org/licenses/>.

The full license can be found in [`LICENSE.txt`](LICENSE.txt).


CHANGELOG.md
# Changelog

## [1.3.0] - UNRELEASED

## [1.2.0] - 2023-05-18

### Changed
- Raised dependencies

## [1.1.0] - 2023-02-25

### Changed
- Upgraded to Debian 11
- Raised dependencies

## [1.0.0] - 2022-01-21

### Changed
- The registration route answers with HTTP 401 if registration invitation code was wrong

## [1.0.0-Preview1] - 2021-02-02

### Added
- Initial release


test_passbutlerserver.py
from flask_testing import TestCase
from passbutlerserver import create_app, db
from passbutlerserver import User

class PassButlerTestCase(TestCase):

    SQLALCHEMY_DATABASE_URI = "sqlite://"
    SQLALCHEMY_TRACK_MODIFICATIONS = False
    TESTING = True

    def create_app(self):
        app = create_app(self)
        return app

    def setUp(self):
        db.create_all()

    def tearDown(self):
        db.session.remove()
        db.drop_all()

class UserTests(PassButlerTestCase):

    def test_get_users(self):
        username = "testuser"
        masterKeyDerivationInformation = "abc"
        masterEncryptionKey = "b"
        itemEncryptionPublicKey = "c"
        itemEncryptionSecretKey = "d"
        settings = "e"
        deleted = False
        modified = 12345678902
        created = 12345678901

        user = User(
            username = username,
            masterKeyDerivationInformation = masterKeyDerivationInformation,
            masterEncryptionKey = masterEncryptionKey,
            itemEncryptionPublicKey = itemEncryptionPublicKey,
            itemEncryptionSecretKey = itemEncryptionSecretKey,
            settings = settings,
            deleted = deleted,
            modified = modified,
            created = created
        )

        db.session.add(user)
        db.session.commit()

        response = self.client.get("/users")

        assert response.status_code == 200

        ## TODO: json order
        assert b'[{"username":"testuser","created":12345678901,"deleted":false,"itemEncryptionPublicKey":"c","modified":12345678902}]' in response.data



# @pytest.fixture
# def client():
#     testDatabaseFileHandle, testDatabaseFileName = tempfile.mkstemp()

#     passbutlerserver.app.config['DATABASE'] = testDatabaseFileName
#     passbutlerserver.app.config['TESTING'] = True
#     client = passbutlerserver.app.test_client()

#     with passbutlerserver.app.app_context():
#         passbutlerserver.initializeDatabase()

#     yield client

#     os.close(testDatabaseFileHandle)
#     os.unlink(passbutlerserver.app.config['DATABASE'])

# def test_empty_users(client):
#     request = client.get('/users')
#     assert b'[]' in request.data

# def test_create_user(client):

#     username = "testuser",
#     masterKeyDerivationInformation = "a",
#     masterEncryptionKey = "b",
#     itemEncryptionPublicKey = "c",
#     itemEncryptionSecretKey = "d",
#     settings = "e",
#     deleted = False,
#     modified = 12345678902,
#     created = 12345678901,

#     client.post('/users', data=dict(
#         username = username,
#         masterKeyDerivationInformation = masterKeyDerivationInformation,
#         masterEncryptionKey = masterEncryptionKey,
#         itemEncryptionPublicKey = itemEncryptionPublicKey,
#         itemEncryptionSecretKey = itemEncryptionSecretKey,
#         settings = settings,
#         deleted = deleted,
#         modified = modified,
#         created = created
#     ))

#     request = client.get('/users')
#     assert b'[]' in request.data

# # assert client.get('/create').status_code == 200
# # client.post('/create', data={'title': 'created', 'body': ''})


passbutler-api.py
#!/usr/bin/env python3

from flask import Flask, request, jsonify, abort, make_response
from flask_marshmallow import Marshmallow
from flask_sqlalchemy import SQLAlchemy
from sqlalchemy.schema import CreateSchema
import json
import os
import time

app = Flask(__name__)
baseDirectory = os.path.abspath(os.path.dirname(__file__))
databasePath = os.path.join(baseDirectory, 'passbutler.sqlite')

app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///' + databasePath
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False

db = SQLAlchemy(app)
ma = Marshmallow(app)

def createJavaDateTimestamp():
    return int(time.time() * 1000)

class User(db.Model):
    username = db.Column(db.String(80), primary_key=True)
    lockTimeout = db.Column(db.Integer)
    deleted = db.Column(db.Boolean)
    modified = db.Column(db.Integer)
    created = db.Column(db.Integer)

    def __init__(self, username, lockTimeout):
        self.username = username
        self.lockTimeout = lockTimeout
        self.deleted = False

        currentTimestamp = createJavaDateTimestamp()
        self.modified = currentTimestamp
        self.created = currentTimestamp

class UserSchema(ma.Schema):
    class Meta:
        fields = ('username', 'lockTimeout', 'deleted', 'modified', 'created')

userSchema = UserSchema()
usersSchema = UserSchema(many=True)

@app.errorhandler(404)
def not_found(error):
    return make_response(jsonify({'error': 'Not found'}), 404)

@app.route("/users", methods=["GET"])
def get_users():
    allUsers = User.query.all()
    result = usersSchema.dump(allUsers)
    return jsonify(result.data)

@app.route("/users", methods=["POST"])
def add_users():

    if (not request.json or not len(request.json) > 0):
        abort(400)

    users = request.json

    for user in users:
        addedUser = User(user['username'], user['lockTimeout'])
        db.session.add(addedUser)

    db.session.commit()

    return ('', 204)

@app.route("/users", methods=["PUT"])
def update_users():

    if (not request.json or not len(request.json) > 0):
        abort(400)

    users = request.json

    for user in users:
        updatedUser = User.query.get(user['username'])

        if (updatedUser is None):
            abort(404)

        ## Update only allowed fields
        updatedUser.lockTimeout = user['lockTimeout']
        updatedUser.modified = user['modified']

    db.session.commit()

    return ('', 204)

@app.route("/user/<username>", methods=["GET"])
def user_detail(username):
    user = User.query.get(username)
    return userSchema.jsonify(user)

## TODO: Check only the user itself is allowed to call this route
@app.route("/user/<username>", methods=["PUT"])
def user_update(username):
    user = User.query.get(username)

    user.lockTimeout = request.json['lockTimeout']
    user.deleted = request.json['deleted']
    user.modified = request.json['modified']

    db.session.commit()
    return userSchema.jsonify(user)

if __name__ == '__main__':
    app.run(host='0.0.0.0', debug=True)


passbutler-server.py
#!/usr/bin/env python3

from flask import Flask, request, jsonify, abort, make_response
from flask_marshmallow import Marshmallow
from flask_sqlalchemy import SQLAlchemy
from sqlalchemy.schema import CreateSchema
import json
import os
import time

app = Flask(__name__)
baseDirectory = os.path.abspath(os.path.dirname(__file__))
databasePath = os.path.join(baseDirectory, 'passbutler.sqlite')

app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///' + databasePath
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False

#CREATE TABLE IF NOT EXISTS `user` (`username` TEXT NOT NULL, `lockTimeout` INTEGER NOT NULL, `deleted` INTEGER NOT NULL, `modified` INTEGER NOT NULL, `created` INTEGER NOT NULL, PRIMARY KEY(`username`));
# INSERT INTO `user` (`username`, `lockTimeout`, `deleted`, `modified`, `created`) VALUES ('a@sicherheitskritisch.de', 1, 0, 0,0);
# INSERT INTO `user` (`username`, `lockTimeout`, `deleted`, `modified`, `created`) VALUES ('b@sicherheitskritisch.de', 1, 0, 0,0);

#curl -X PUT -H "Content-Type: application/json" -d '{"deleted":0,"lockTimeout":3,"modified":1}' "http://127.0.0.1:5000/user/a@sicherheitskritisch.de"


db = SQLAlchemy(app)
ma = Marshmallow(app)

def createJavaDateTimestamp():
    return int(time.time() * 1000)

class User(db.Model):
    username = db.Column(db.String(80), primary_key=True)
    lockTimeout = db.Column(db.Integer)
    deleted = db.Column(db.Boolean)
    modified = db.Column(db.Integer)
    created = db.Column(db.Integer)

    def __init__(self, username, lockTimeout):
        self.username = username
        self.lockTimeout = lockTimeout
        self.deleted = False

        currentTimestamp = createJavaDateTimestamp()
        self.modified = currentTimestamp
        self.created = currentTimestamp

class UserSchema(ma.Schema):
    class Meta:
        fields = ('username', 'lockTimeout', 'deleted', 'modified', 'created')

userSchema = UserSchema()
usersSchema = UserSchema(many=True)

@app.errorhandler(404)
def not_found(error):
    return make_response(jsonify({'error': 'Not found'}), 404)

@app.route("/users", methods=["GET"])
def get_users():
    allUsers = User.query.all()
    result = usersSchema.dump(allUsers)
    return jsonify(result.data)

@app.route("/users", methods=["POST"])
def add_users():

    if (not request.json or not len(request.json) > 0):
        abort(400)

    users = request.json

    for user in users:
        addedUser = User(user['username'], user['lockTimeout'])
        db.session.add(addedUser)

    db.session.commit()

    return ('', 204)

@app.route("/users", methods=["PUT"])
def update_users():

    if (not request.json or not len(request.json) > 0):
        abort(400)

    users = request.json

    for user in users:
        updatedUser = User.query.get(user['username'])

        if (updatedUser is None):
            abort(404)

        ## Update only allowed fields
        updatedUser.lockTimeout = user['lockTimeout']
        updatedUser.modified = user['modified']
        updatedUser.deleted = user['deleted']

    db.session.commit()

    return ('', 204)

@app.route("/user/<username>", methods=["GET"])
def user_detail(username):
    user = User.query.get(username)
    return userSchema.jsonify(user)

## TODO: Check only the user itself is allowed to call this route
@app.route("/user/<username>", methods=["PUT"])
def user_update(username):
    user = User.query.get(username)

    user.lockTimeout = request.json['lockTimeout']
    user.deleted = request.json['deleted']
    user.modified = request.json['modified']

    db.session.commit()
    return userSchema.jsonify(user)

if __name__ == '__main__':
    app.run(host='0.0.0.0', debug=True)
